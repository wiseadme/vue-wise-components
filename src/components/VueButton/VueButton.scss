.vue-button {
  @include flex(center, center);
  position: relative;
  height: 40px;
  min-width: 80px;
  font-size: 0.8em;
  text-transform: uppercase;
  cursor: pointer;
  border-radius: 3px;
  border: none;
  outline: none;
  box-shadow: 0 0 4px 1px rgba(0, 0, 0, 0.5);
  transition: box-shadow 0.3s ease;
  overflow: hidden;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;

  &:after {
    content: "";
    display: block;
    position: absolute;
    z-index: 1;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 0;
    background: rgba(255, 255, 255, 0.3);
    transition: all 0.3s ease;
  }

  &:before {
    content: "";
    display: block;
    position: absolute;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: rgba(0, 0, 0, .2);
    opacity: 0;
    transition: opacity 0.3s ease;
  }

  &:hover {
    box-shadow: 0 0 12px 0 rgba(0, 0, 0, 0.5);

    &:active {
      box-shadow: none;
      transition: all .3s;

      &:before {
        opacity: 1;
      }
    }
  }

  &:hover:after {
    height: 100%;
    opacity: 0;
  }

  &--danger {
    background: $danger;
  }

  &--success {
    background: $success;
  }

  &--primary {
    background: $primary;
  }

  &--warning {
    background-color: $warning;
  }

  &--disabled {
    background: $disabled;
    box-shadow: none;

    &:hover {
      box-shadow: none;

      &:after {
        height: 0;
      }
    }
  }

  &__text {
    color: #FFFFFF;
    margin: 0 2.5px;
  }
  &__content {
    margin: 0 2.5px;
  }
}
