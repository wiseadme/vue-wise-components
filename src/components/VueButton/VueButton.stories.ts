import { storiesOf } from '@storybook/vue'
import { action } from '@storybook/addon-actions'
import { withKnobs, text } from '@storybook/addon-knobs'

const story = storiesOf('VueButton', module)

story.addDecorator(withKnobs)

story.add('primary', () => ({
  methods: {
    addClick: () => action('clicked')
  },
  props: {
    text: {
      default: text('btn-label', 'stop')
    }
  },
  template: `
    <div class="btn-wrap">
      <vue-button type="primary" :text="text" @click="addClick">some icon</vue-button>
    </div>`
}))
  .add('success', () => ({
    template: `
      <div class="btn-wrap">
        <vue-button type="success" text="send"></vue-button>
      </div>`
  }))
  .add('warning', () => ({
    template: `
      <div class="btn-wrap">
        <vue-button type="warning" text="send"></vue-button>
      </div>`
  }))
