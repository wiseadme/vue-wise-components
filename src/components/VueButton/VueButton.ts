import './VueButton.scss'

import Vue from 'vue'
import { VNode, VNodeData } from 'vue/types'

import { getSlot } from '@/helpers'

const buttonTypes: string[] = [
  'primary',
  'success',
  'warning',
  'danger'
]

export default Vue.extend({
  name: 'VueButton',
  props: {
    text: String,
    type: {
      type: String,
      validator(value: string): boolean {
        return buttonTypes.includes(value)
      }
    },
    color: String,
    disabled: Boolean
  },

  computed: {
    content(): VNode {
      return this.$createElement('span',
        {
          staticClass: 'vue-button__text',
        }, this.text)
    },

    slotContent(): VNode | null {
      const slotContent = getSlot(this)
      if (slotContent) {
        return this.$createElement('div', {
          staticClass: 'vue-button__content'
        }, slotContent)
      }
      return null
    },

    renderData(): VNodeData {
      return {
        staticClass: 'vue-button',
        class: {
          ...this.classes
        },
        style: {
          backgroundColor: this.color
        },
        on: {
          click: () => {
            if (this.disabled) return
            this.$emit('click')
          }
        }
      }
    },

    classes(): Record<string, boolean> {
      return {
        [`vue-button--${ this.type }`]: !this.color && !this.disabled,
        'vue-button--disabled': this.disabled,
        [this.color]: !!this.color
      }
    }
  },

  render(h): VNode {
    if (this.slotContent) {
      this.content.children!.push(this.slotContent!)
    }
    const content = [this.content]
    return h('button', this.renderData, content)
  }
})
