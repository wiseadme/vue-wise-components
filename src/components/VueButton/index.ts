import VueButton from './VueButton'

export default VueButton
