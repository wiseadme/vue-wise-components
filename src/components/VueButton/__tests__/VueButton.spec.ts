import { shallowMount } from '@vue/test-utils'
import VueButton from '../VueButton'

describe('VueButton.ts', () => {
  it('renders props.text if it passed', () => {
    const text = 'button'
    const wrapper = shallowMount(VueButton, {
      propsData: { text }
    })
    expect(wrapper.text()).toMatch(text)
  })

  it('renders button type by props.type', () => {
    const type = 'primary'
    const wrapper = shallowMount(VueButton, {
      propsData: { type }
    })
    expect(wrapper.props().type).toEqual(type)
  })

  it('make snapshot', () => {
    const wrapper = shallowMount(VueButton, {
      propsData: {
        type: 'success',
        text: 'send'
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})

