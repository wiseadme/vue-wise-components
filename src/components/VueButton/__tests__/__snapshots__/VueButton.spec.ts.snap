// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VueButton.ts make snapshot 1`] = `
<button
  class="vue-button vue-button--success"
>
  <span
    class="vue-button__text"
  >
    send
  </span>
</button>
`;
